pkg/tfbridge/info.go

```diff
@@ -888,29 +888,64 @@ func GetModuleMajorVersion(version string) string {
 }
 
 // MakeMember manufactures a type token for the package and the given module and type.
+//
+// Depreciated: Use MakeResource or call into the `tokens` module in
+// "github.com/pulumi/pulumi/sdk/v3/go/common/tokens" directly.
 func MakeMember(pkg string, mod string, mem string) tokens.ModuleMember {
 	return tokens.ModuleMember(pkg + ":" + mod + ":" + mem)
 }
 
 // MakeType manufactures a type token for the package and the given module and type.
+//
+// Depreciated: Use MakeResource or call into the `tokens` module in
+// "github.com/pulumi/pulumi/sdk/v3/go/common/tokens" directly.
 func MakeType(pkg string, mod string, typ string) tokens.Type {
 	return tokens.Type(MakeMember(pkg, mod, typ))
 }
 
 // MakeDataSource manufactures a standard Pulumi function token given a package, module, and data source name.  It
 // automatically uses the main package and names the file by simply lower casing the data source's
 // first character.
-func MakeDataSource(pkg string, mod string, res string) tokens.ModuleMember {
-	fn := string(unicode.ToLower(rune(res[0]))) + res[1:]
-	return MakeMember(pkg, mod+"/"+fn, res)
+//
+// Invalid inputs panic.
+func MakeDataSource(pkg string, mod string, name string) tokens.ModuleMember {
+	if !tokens.IsName(name) {
+		panic(fmt.Errorf("invalid datasource name: '%s'", name))
+	}
+	modT := makeModule(pkg, mod, name)
+	return tokens.NewModuleMemberToken(modT, tokens.ModuleMemberName(name))
+}
+
+// makeModule manufactures a standard pulumi module from a (pkg, mod, member) triple.
+//
+// For example:
+//
+//	(pkg, mod, Resource) => pkg:mod/resource
+//
+// Invalid inputs panic.
+func makeModule(pkg, mod, member string) tokens.Module {
+	mod += "/" + string(unicode.ToLower(rune(member[0]))) + member[1:]
+	if !tokens.IsQName(pkg) {
+		panic(fmt.Errorf("invalid pkg name: '%s'", pkg))
+	}
+	pkgT := tokens.NewPackageToken(tokens.PackageName(pkg))
+	if !tokens.IsQName(mod) {
+		panic(fmt.Errorf("invalid module name: '%s'", mod))
```

Comment:
It's a bit ruthless to panic on invalid inputs but we can't change the signature to `(T, err)` in a non-breaking way so that's probably an improvement overall, justified move.

Reply:
Non-compliant tokens will fail during schema checks later down the line. I agree it's ruthless, but at least it won't break any existing providers. It also matches the way the `tokens` module is written (panicking on invalid inputs).

Reply:
Yes. LGTM - good move for now.